File: server/consume.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <sstream>
#include <diaspora/Driver.hpp>
#include <diaspora/TopicHandle.hpp>
#include <diaspora/Consumer.hpp>

int main(int argc, char** argv)
{
    std::string group_file = argv[1];
    std::string topic_name = (argc > 2) ? argv[2] : "darshan";
    int want = (argc > 3) ? std::atoi(argv[3]) : 10;

    std::ifstream f(group_file);
    std::stringstream ss; ss << f.rdbuf();
    std::string opts = std::string("{\"group_file\":\"") + group_file + "\"}";

    auto driver = diaspora::Driver::New("mofka", diaspora::Metadata{opts});
    auto topic  = driver.openTopic(topic_name);
    auto consumer = topic.consumer("darshan-consumer", driver.defaultThreadPool());

    for (int i = 0; i < want; i++) {
        auto event = consumer.pull().wait(10000);
        if (!event) { std::cout << "(timeout waiting for event " << i << ")\n"; break; }
        auto& doc = event->metadata().json();
        std::cout << "[" << i << "] module=" << doc.value("module", "?")
                  << " op=" << doc.value("op", "?")
                  << " file=" << doc.value("file", "?").substr(0, 48)
                  << " len=" << doc.value("len", (int64_t)-1) << "\n";
    }
    std::cout << "--- done consuming topic '" << topic_name << "' ---\n";
    return 0;
}

File: server/consume.py
import sys, json
# Current mochi.mofka API: AdaptiveBatchSize lives in mochi.mofka.client
# (older harness imported it from pymofka_client). Events from the darshan
# connector are metadata-only, so we use a no-op data selector/allocator.
from mochi.mofka.client import MofkaDriver, AdaptiveBatchSize

groupfile = sys.argv[1]
topic = sys.argv[2] if len(sys.argv) > 2 else "darshan"
want = int(sys.argv[3]) if len(sys.argv) > 3 else 10

driver = MofkaDriver(group_file=groupfile)
th = driver.open_topic(topic)
consumer = th.consumer(
    "darshan-consumer",
    batch_size=AdaptiveBatchSize,
    data_selector=lambda metadata, descriptor: None,  # metadata-only: load no data
    data_allocator=lambda metadata, descriptor: [],
)

n = 0
while n < want:
    ev = consumer.pull().wait()
    # newer bindings hand back metadata already parsed into a dict; older ones a JSON string
    md = ev.metadata if isinstance(ev.metadata, dict) else json.loads(ev.metadata)
    print(f"[{n}] module={md.get('module')} op={md.get('op')} "
          f"rank={md.get('rank')} file={md.get('file','')[:50]} bytes={md.get('len')}")
    n += 1
print(f"--- received {n} events from topic '{topic}' ---")
